strawberryfields/engine.py

```diff
@@ -277,15 +272,57 @@ def _broadcast_nones(val, dim):
                 p = p.compile(target, **compile_options)
             p.lock()
 
-            self._run_program(p, **kwargs)
-            shots = kwargs.get("shots", 1)
-            self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
+            _, values = self._run_program(p, **kwargs)
             self.run_progs.append(p)
 
+            # check number of measurments and combine samples if needed
+            if len(values) > 1:
+                self._combine_samples(values, p)
+            elif len(values) == 1:
+                self.samples = values[0]
+            else:  # if no measurement was made
+                self.samples = np.array([[]])
+
             prev = p
 
-        if self.samples is not None:
-            return Result(np.array(self.samples).T)
+        return Result(self.samples)
+
+    def _combine_samples(self, values, program):
+        """Combines samples if multiple measurements were made.
+
+        If batches are used, the samples will be fed to ``Result.combine_samples``
+        one-batch-at-a-time, which combines the samples and returns them with shape
+        ``(batches, shots, measured_modes)``.
+
+        Args:
+            values (list[array, tensor]): the sample measurements in a list
+            program (Program): quantum program that is run
+        """
+        # get nested order of measured modes
+        sort_order = [
+            np.array([r.ind for r in c.reg]) for c in program.circuit if "Measure" in c.op.__str__()
+        ]
+
+        # check for duplicate mode-measures
+        if len(np.hstack(sort_order)) != len(set(np.hstack(sort_order))):
+            raise RuntimeError("Modes can only be measured once inside a circuit.")
+
+        # pylint: disable=import-outside-toplevel
+        if self.backend_name == "tf":
+            from tensorflow import convert_to_tensor
+
+            if self.backend_options.get("batch_size", 0):  # if batches are used
+                self.samples = []
+                for i in range(self.backend_options.get("batch_size", 0)):
+                    # choose a single batch from values and combine the samples; do for all batches
+                    single_batch_values = [v[i] for v in values]
+                    single_batch_samples = Result.combine_samples(single_batch_values, sort_order)
+                    self.samples.append(single_batch_samples)
```

Comment:
Just to double check what is happening here, we are essentially 'vectorizing' (in the loose sense of the word, since this is a for-loop) the `combine_samples()` function for each batch dimension?